### PR TITLE
Consolidate manifest writing for Turbopack in a single function

### PR DESCRIPTION
**Pull request by timneutkens**
## What?

I'm working on consolidating a bunch of the file writing related pieces in the Turbopack handling in the dev server so that it can be abstracted out as it's needed for `next build` too.

These changes make sure that there is a single `writeManifests()` instead of picking specific manifests to write.

We can optimize this later but for now the overhead of writing them to disk separately is negligible.

Closes NEXT-1884

### CODE REVIEW

**Comment by timneutkens** [comment]
Current dependencies on/for this PR:
* `canary`
  * **PR #59663** <a href="https://app.graphite.dev/github/pr/vercel/next.js/59663?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/vercel/next.js/59663?utm_source=stack-comment).

File: packages/next/src/server/lib/router-utils/setup-dev-bundler.ts [hunk 1]
```diff
@@ -552,7 +552,7 @@ async function startWatcher(opts: SetupOpts) {
         hmrBuilding = true
       }
       hmrPayloads.set(`${key}:${id}`, payload)
-      hmrEventHappend = true
+      hmrEventHappened = true
```
Comment (timneutkens) [inline packages/next/src/server/lib/router-utils/setup-dev-bundler.ts:555]: Noticed a typo here, corrected it.

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 12-15-Consolidate_manifest_writing_for_Turbopack_in_a_single_function | Change |
| - | - | - | - |
| buildDuration | 12.8s | 12.9s | ⚠️ +150ms |
| buildDurationCached | 7s | 6.1s | N/A |
| nodeModulesSize | 200 MB | 200 MB | N/A |
| nextStartRea..uration (ms) | 420ms | 428ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 12-15-Consolidate_manifest_writing_for_Turbopack_in_a_single_function | Change |
| - | - | - | - |
| 170-HASH.js gzip | 26.8 kB | 26.8 kB | N/A |
| 199.HASH.js gzip | 181 B | 185 B | N/A |
| 3f784ff6-HASH.js gzip | 53.3 kB | 53.3 kB | ✓ |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 241 B | 241 B | ✓ |
| main-HASH.js gzip | 31.7 kB | 31.7 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | N/A |
| Overall change | 98.7 kB | 98.7 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 12-15-Consolidate_manifest_writing_for_Turbopack_in_a_single_function | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 12-15-Consolidate_manifest_writing_for_Turbopack_in_a_single_function | Change |
| - | - | - | - |
| _app-HASH.js gzip | 195 B | 194 B | N/A |
| _error-HASH.js gzip | 183 B | 182 B | N/A |
| amp-HASH.js gzip | 501 B | 501 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 255 B | 255 B | ✓ |
| head-HASH.js gzip | 349 B | 350 B | N/A |
| hooks-HASH.js gzip | 368 B | 369 B | N/A |
| image-HASH.js gzip | 4.27 kB | 4.27 kB | N/A |
| index-HASH.js gzip | 255 B | 256 B | N/A |
| link-HASH.js gzip | 2.61 kB | 2.6 kB | N/A |
| routerDirect..HASH.js gzip | 311 B | 309 B | N/A |
| script-HASH.js gzip | 384 B | 384 B | ✓ |
| withRouter-HASH.js gzip | 307 B | 306 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 1.57 kB | 1.57 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 12-15-Consolidate_manifest_writing_for_Turbopack_in_a_single_function | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 482 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 12-15-Consolidate_manifest_writing_for_Turbopack_in_a_single_function | Change |
| - | - | - | - |
| index.html gzip | 528 B | 529 B | N/A |
| link.html gzip | 542 B | 541 B | N/A |
| withRouter.html gzip | 524 B | 524 B | ✓ |
| Overall change | 524 B | 524 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 12-15-Consolidate_manifest_writing_for_Turbopack_in_a_single_function | Change |
| - | - | - | - |
| edge-ssr.js gzip | 93.7 kB | 93.7 kB | N/A |
| page.js gzip | 146 kB | 146 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 12-15-Consolidate_manifest_writing_for_Turbopack_in_a_single_function | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 627 B | 623 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 37.4 kB | 37.4 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 2.07 kB | 2.07 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js 12-15-Consolidate_manifest_writing_for_Turbopack_in_a_single_function | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 168 kB | 168 kB | ✓ |
| app-page-exp..prod.js gzip | 94.1 kB | 94.1 kB | ✓ |
| app-page-tur..prod.js gzip | 94.8 kB | 94.8 kB | ✓ |
| app-page-tur..prod.js gzip | 89.4 kB | 89.4 kB | ✓ |
| app-page.run...dev.js gzip | 138 kB | 138 kB | ✓ |
| app-page.run..prod.js gzip | 88.7 kB | 88.7 kB | ✓ |
| app-route-ex...dev.js gzip | 24 kB | 24 kB | ✓ |
| app-route-ex..prod.js gzip | 16.6 kB | 16.6 kB | ✓ |
| app-route-tu..prod.js gzip | 16.6 kB | 16.6 kB | ✓ |
| app-route-tu..prod.js gzip | 16.2 kB | 16.2 kB | ✓ |
| app-route.ru...dev.js gzip | 23.4 kB | 23.4 kB | ✓ |
| app-route.ru..prod.js gzip | 16.2 kB | 16.2 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.64 kB | 9.64 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-turbo...prod.js gzip | 21.9 kB | 21.9 kB | ✓ |
| pages.runtim...dev.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim..prod.js gzip | 21.9 kB | 21.9 kB | ✓ |
| server.runti..prod.js gzip | 49.4 kB | 49.4 kB | ✓ |
| Overall change | 930 kB | 930 kB | ✓ |

</details>

</details>
Commit: f3a22871bcc29957419d64a44c5dbac4b3d10632

File: packages/next/src/server/lib/router-utils/setup-dev-bundler.ts [hunk 2]
```diff
@@ -1041,6 +1041,18 @@ async function startWatcher(opts: SetupOpts) {
       )
     }
 
+    async function writeManifests(): Promise<void> {
```
Comment (feedthejim) [inline packages/next/src/server/lib/router-utils/setup-dev-bundler.ts:1044]: Is This A Pigeon

**Review by feedthejim — APPROVED** [review]

File: packages/next/src/server/lib/router-utils/setup-dev-bundler.ts [hunk 3]
```diff
@@ -1041,6 +1041,18 @@ async function startWatcher(opts: SetupOpts) {
...
+      await writeBuildManifest(opts.fsChecker.rewrites)
+      await writeAppBuildManifest()
+      await writePagesManifest()
+      await writeAppPathsManifest()
+      await writeMiddlewareManifest()
+      await writeActionManifest()
+      await writeFontManifest()
+      await writeLoadableManifest()
+      await writeFallbackBuildManifest()
```
Comment (gnoff) [inline packages/next/src/server/lib/router-utils/setup-dev-bundler.ts:1053]: do we want to parallelize these? I have not idea if they have side effects so maybe not safe but seemingly the io parts could be done in parallel if they do not interact

File: packages/next/src/server/lib/router-utils/setup-dev-bundler.ts [hunk 4]
```diff
@@ -1166,6 +1178,8 @@ async function startWatcher(opts: SetupOpts) {
               'edge'
             )
             await loadMiddlewareManifest('instrumentation', 'instrumentation')
+            await writeManifests()
```
Comment (timneutkens) [inline packages/next/src/server/lib/router-utils/setup-dev-bundler.ts:1181]: Noticed this bug where it didn't write the manifest with instrumentation

**Review by gnoff — APPROVED** [review]

File: packages/next/src/server/lib/router-utils/setup-dev-bundler.ts (hunk 3 above)
Comment (timneutkens) [reply packages/next/src/server/lib/router-utils/setup-dev-bundler.ts:1053]: Overall we'll likely want to make these sync fs actually, but kept it as-is for now, can optimize later 👍

**Pull request merged**